DEVDOCS-3406b: GraphQL-related scope creep (#1496)
bigcommerce/dev-docs: docs/api-docs/storefront/graphql/graphql-api-overview.md

<!-- theme: info -->
> #### Note

> > * GraphQL Playground is a [GraphQL IDE](https://github.com/andev-software/graphql-ide) built on Electron. For more information, see [GraphQL Playground](https://electronjs.org/apps/graphql-playground) on [electronjs.org](https://electronjs.org)

Author: agree. also it's being deprecated. axed in a coming push.

## Authentication

GraphQL Storefront API requests are authenticated with tokens sent using the HTTP `Authorization` header.

> ```shell title="Example Authorization header using cURL"

Reviewer: ```suggestion
```shell title="Example authorization header using cURL"

```

Author: @bc-andreadao I'm treating `Authorization` as the proper name of the `Authorization` header because it's a standard header, not just a BC one.

### Creating a token

> Use the [Create a GraphQL API token](/api-reference/storefront/graphql-api-tokens/api-token/createtoken) REST endpoint to request JWT bearer tokens that authenticate cross-origin requests to the Storefront API. Your JWT will have the same OAuth scopes as the REST API account that creates it, so ensure that the API account you use for the following request has sufficient permissions to make your GraphQL queries. 

Author: not technically per the [OAuth guardians](https://oauth.net/2/bearer-tokens/#:~:text=Bearer%20Tokens%20are%20the%20predominant,such%20as%20JSON%20Web%20Tokens.)... but revised a bit like so:

```suggestion
Use the [Create a GraphQL API token](/api-reference/storefront/graphql-api-tokens/api-token/createtoken) REST endpoint to request JWT-style bearer tokens that authenticate cross-origin requests to the Storefront API. Your JWT will have the same OAuth scopes as the REST API account that creates it, so ensure that the API account you use for the following request has sufficient permissions to make your GraphQL queries. 

```

Customer impersonation token authenticated requests made to the GraphQL API receive store information from the perspective of the customer corresponding to the customer ID specified in the `X-Bc-Customer-Id` header sent with the GraphQL `POST` request. Pricing, product availability, customer account, and customer details will be reflected.

> Customer impersonation tokens should **never** be exposed publicly, for example, to JavaScript or HTML. These tokens should not be used for frontend requests.

Author: agree, combined in a coming push.

> Unlike normal GraphQL Storefront API tokens, they are sensitive and should be treated like secrets, just as you might treat an OAuth access token for BigCommerce's REST APIs. Attempts to run requests using these tokens from a web browser will be rejected.

Author: combined per above.

GraphQL Storefront API calls can be made directly from within a Stencil theme or from a script in [Storefront > Script Manager](https://support.bigcommerce.com/s/article/Using-Script-Manager).

> Here's an example request using the `{{settings.storefront_api.token}}` handlebars object and [JavaScript's Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API):

Author: love it. done. preview:

```suggestion
The following example request uses the `{{settings.storefront_api.token}}` handlebars object and [JavaScript's Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch):
```

```js title="Example request: GraphQL query using Stencil token" lineNumbers
fetch('/graphql', {
  method: 'POST',
  credentials: 'same-origin',
  headers: {
    'Content-Type': 'application/json',
    'Authorization': 'Bearer {{ settings.storefront_api.token }}'
  },
  body: JSON.stringify({
    query: `query MyFirstQuery {
      site {
        settings {
          storeName
        }
        products {
          edges {
            node {
              name
              sku
              prices {
                retailPrice {
                  value
                  currencyCode
                }
                price {
                  value
                  currencyCode
                }
              }
            }
          }
        }
      }
    }`
  })
})
.then(res => console.log(res))
.catch(error => console.error(error));
```

> In addition to JavaScript's native Fetch API, you can use GraphQL a client library to query BigCommerce. [Apollo](https://www.apollographql.com/docs/react/) is a popular GraphQL client that's easy to use in BigCommerce themes. For a quick example of adding Apollo Client to Cornerstone, checkout this [Cornerstone commit](https://github.com/bigcommerce/cornerstone/commit/508feeb1b00d2bb2940771e5e91250a08b6be4d9) on GitHub. GraphQL is a standard with client libraries in many languages, so feel free to explore your options.

Author: I like it and ran with it:

```suggestion
Client libraries like [Apollo](https://www.apollographql.com/docs/react/) offer features that can simplify GraphQL implementations, such as [declarative data fetching](https://www.apollographql.com/docs/react/data/queries), [state management](https://www.apollographql.com/docs/react/local-state/local-state-management), and [caching](https://www.apollographql.com/docs/react/caching/overview) for more consistent UI components. For an example of adding Apollo Client to the Cornerstone theme, check out this [Cornerstone commit](https://github.com/bigcommerce/cornerstone/commit/508feeb1b00d2bb2940771e5e91250a08b6be4d9) on GitHub.

```

You have two options for public URLs you can use to run requests:

- The store's default storefront URL, e.g. `https://store.com/graphql`
- The store's permanent URL, e.g. `https://store-STOREHASH.mybigcommerce.com/graphql`

> #### I want to run requests from the perspective of another Channel

Author: done

> You should use the Channel's Permanent URL of the form `https://store-STOREHASH-CHANNELID.mybigcommerce.com/graphql`.

Author: done

> For example, if your store hash is `abc123` and your channel ID is `456`, the correct URL would be `https://store-abc123-456.mybigcommerce.com/graphql`.

Author: done

> Note that you must create your Storefront API Token with the same channel ID, or your request will be rejected.

Reviewer 2: ```suggestion
You must create your storefront API token with the same channel ID, or the server will reject your request.
```

Author: converted to callout; improved by incorporating this suggestion.

> In order for the Channel's Permanent URL to be available, you must [create a Site](/api-reference/cart-checkout/sites-routes-api/sites/post-site) for the channel.

Reviewer 2: ```suggestion
You must [create a site](/api-reference/cart-checkout/sites-routes-api/sites/post-site) for the channel for the channel's permanent URL to be available.

```

Author: yep

You can run this query against an example storefront using the [GraphQL Playground](/graphql-playground?tabs=firstThreeProducts).

> The results will look something like the following:

Reviewer 2: ```suggestion
The results look something like the following:

```

Author: yes

```

> This same approach can be used to *slice* any GraphQL connection and paginate through the *slices* using `startCursor` and `endCursor`. For example, we could get the first thirty brands with the following query:

Reviewer 2: ```suggestion
This same approach can be used to *slice* any GraphQL connection and paginate through the *slices* using `startCursor` and `endCursor`. For example, the following query gets the first thirty brands:

```

Author: done

        ]
      }
    }
  }
}
```

> the next thirty could be retrieved by making a new query and passing in the `endCursor` from the first page of results:

Reviewer 2: ```suggestion
You can retrieve the next thirty by making a new query and passing in the `endCursor` from the first page of results:

```

Author: done
